File: generator.py
"""
AgentQL code generator for Legacy-to-AgentQL migration.

Turns a batch of ``TranslatedLocator`` records into an AgentQL query
string and a runnable ``sync_playwright`` script that wraps the page,
calls ``query_elements``, and replays the original interactions.
"""

from __future__ import annotations

import ast
from dataclasses import dataclass
from typing import Sequence

from parser import Interaction
from translator import TranslatedLocator, is_valid_agentql_name

DEFAULT_URL = "https://example.com"
DEFAULT_FUNCTION_NAME = "run_migrated_flow"

_INTERACTION_METHODS: dict[Interaction, str] = {
    Interaction.CLICK: "click",
    Interaction.FILL: "fill",
    Interaction.TYPE: "type",
    Interaction.CHECK: "check",
    Interaction.SELECT: "select_option",
    Interaction.HOVER: "hover",
    Interaction.WAIT: "wait_for",
}

_VALUE_INTERACTIONS = frozenset(
    {
        Interaction.FILL,
        Interaction.TYPE,
        Interaction.SELECT,
    }
)


@dataclass(frozen=True, slots=True)
class GeneratedAction:
    """One interaction line in the generated AgentQL script.

    Attributes:
        name: AgentQL response field (snake_case).
        interaction: Playwright-style action to invoke.
        fill_value: Literal argument for fill/type/select when known.
        source_lineno: Line in the legacy script this action came from.
    """

    name: str
    interaction: Interaction
    fill_value: str | None = None
    source_lineno: int | None = None


@dataclass(frozen=True, slots=True)
class GeneratedScript:
    """Complete AgentQL migration artifact for one legacy file.

    Attributes:
        query: AgentQL query text passed to ``page.query_elements``.
        python_source: Full generated Python module.
        url: Navigation target used in ``page.goto``.
        field_names: Ordered AgentQL field names in the query.
        actions: Ordered interaction steps after the query.
        source_path: Optional path of the legacy script that was migrated.
    """

    query: str
    python_source: str
    url: str
    field_names: tuple[str, ...]
    actions: tuple[GeneratedAction, ...]
    source_path: str | None = None


def _python_string_literal(value: str) -> str:
    """Render *value* as a safely escaped Python string literal."""
    return repr(value)


def build_agentql_query(translations: Sequence[TranslatedLocator]) -> str:
    """Build an AgentQL query block from translated field names.

    Emits a multi-line query such as::

        {
            user_input
            submit_button
        }

    Duplicate names are kept once (first occurrence wins) so the query
    stays valid even if uniqueness slipped upstream.
    """
    seen: set[str] = set()
    fields: list[str] = []
    for item in translations:
        name = item.name.strip()
        if not name or name in seen:
            continue
        if not is_valid_agentql_name(name):
            continue
        seen.add(name)
        fields.append(name)

    if not fields:
        return "{\n}"

    body = "\n".join(f"    {name}" for name in fields)
    return "{\n" + body + "\n}"


def action_statement(action: GeneratedAction, *, response_var: str = "elements") -> str:
    """Return one Python statement that performs *action* on *response_var*.

    Value-bearing interactions (fill / type / select_option) include a
    safely quoted literal. Unsupported / none interactions become a
    comment so generation never invents a bogus Playwright call.
    """
    method = _INTERACTION_METHODS.get(action.interaction)
    target = f"{response_var}.{action.name}"

    if method is None:
        return (
            f"# skip {action.name}: unsupported interaction "
            f"{action.interaction.value!r}"
        )

    if action.interaction in _VALUE_INTERACTIONS:
        literal = _python_string_literal(action.fill_value or "")
        return f"{target}.{method}({literal})"

    return f"{target}.{method}()"


def generate_script(
    translations: Sequence[TranslatedLocator],
    *,
    url: str = DEFAULT_URL,
    function_name: str = DEFAULT_FUNCTION_NAME,
    headless: bool = True,
    source_path: str | None = None,
) -> GeneratedScript:
    """Emit an AgentQL + sync_playwright script (stub)."""
    raise NotImplementedError("script generator not yet implemented")

File: web/lib/generator.ts
import { isValidAgentqlName } from "./translator";
import type {
  GeneratedAction,
  GeneratedScript,
  Interaction,
  TranslatedLocator,
} from "./types";

export const DEFAULT_URL = "https://example.com";
export const DEFAULT_FUNCTION_NAME = "run_migrated_flow";

const INTERACTION_METHODS: Partial<Record<Interaction, string>> = {
  click: "click",
  fill: "fill",
  type: "type",
  check: "check",
  select_option: "select_option",
  hover: "hover",
  wait_for: "wait_for",
};

const VALUE_INTERACTIONS = new Set<Interaction>(["fill", "type", "select_option"]);

export function pythonRepr(value: string): string {
  const escaped = value
    .replace(/\\/g, "\\\\")
    .replace(/\n/g, "\\n")
    .replace(/\r/g, "\\r")
    .replace(/\t/g, "\\t");
  if (escaped.includes("'") && !escaped.includes('"')) {
    return `"${escaped.replace(/"/g, '\\"')}"`;
  }
  return `'${escaped.replace(/'/g, "\\'")}'`;
}

export function buildAgentqlQuery(translations: TranslatedLocator[]): string {
  const seen = new Set<string>();
  const fields: string[] = [];
  for (const item of translations) {
    const name = item.name.trim();
    if (!name || seen.has(name) || !isValidAgentqlName(name)) continue;
    seen.add(name);
    fields.push(name);
  }
  if (!fields.length) return "{\n}";
  return "{\n" + fields.map((name) => `    ${name}`).join("\n") + "\n}";
}

export function actionStatement(
  action: GeneratedAction,
  responseVar = "elements",
): string {
  const method = INTERACTION_METHODS[action.interaction];
  const target = `${responseVar}.${action.name}`;
  if (!method) {
    return `# skip ${action.name}: unsupported interaction ${pythonRepr(action.interaction)}`;
  }
  if (VALUE_INTERACTIONS.has(action.interaction)) {
    return `${target}.${method}(${pythonRepr(action.fill_value || "")})`;
  }
  return `${target}.${method}()`;
}

export function fieldNames(translations: TranslatedLocator[]): string[] {
  return buildAgentqlQuery(translations)
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line && line !== "{" && line !== "}");
}

export function translationsToActions(
  translations: TranslatedLocator[],
): GeneratedAction[] {
  return translations
    .filter((item) => isValidAgentqlName(item.name))
    .map((item) => ({
      name: item.name,
      interaction: item.locator.interaction,
      fill_value: item.locator.fill_value,
      source_lineno: item.locator.lineno,
    }));
}

function renderPythonModule(opts: {
  query: string;
  actions: GeneratedAction[];
  url: string;
  functionName: string;
  headless: boolean;
  sourcePath: string | null;
}): string {
  const origin = opts.sourcePath || "legacy script";
  const queryLiteral = `"""\n${opts.query}\n"""`;
  const actionLines = opts.actions.map((action) => `        ${actionStatement(action)}`);
  const actionsBlock = actionLines.length
    ? actionLines.join("\n")
    : "        pass  # no interactions discovered";

  return `"""
Auto-generated AgentQL migration of ${origin}.

Regenerate with \`\`python generator.py\`\` — do not hand-edit locators here.
Requires AGENTQL_API_KEY (see .env.example).
"""

import agentql
from playwright.sync_api import sync_playwright

QUERY = ${queryLiteral}


def ${opts.functionName}() -> None:
    """Run the migrated flow with semantic AgentQL selectors."""
    with sync_playwright() as playwright:
        browser = playwright.chromium.launch(headless=${opts.headless ? "True" : "False"})
        page = agentql.wrap(browser.new_page())
        page.goto(${pythonRepr(opts.url)})

        elements = page.query_elements(QUERY)
${actionsBlock}

        browser.close()


if __name__ == "__main__":
    ${opts.functionName}()
`;
}

export function generateScript(
  translations: TranslatedLocator[],
  opts: {
    url?: string;
    functionName?: string;
    headless?: boolean;
    sourcePath?: string | null;
  } = {},
): GeneratedScript {
  const url = opts.url ?? DEFAULT_URL;
  const query = buildAgentqlQuery(translations);
  const names = fieldNames(translations);
  const actions = translationsToActions(translations);
  const python_source = renderPythonModule({
    query,
    actions,
    url,
    functionName: opts.functionName ?? DEFAULT_FUNCTION_NAME,
    headless: opts.headless ?? true,
    sourcePath: opts.sourcePath ?? null,
  });
  return {
    query,
    python_source,
    url,
    field_names: names,
    actions,
    source_path: opts.sourcePath ?? null,
  };
}
